instrumentation/graphql/lib/opentelemetry/instrumentation/graphql/instrumentation.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+# Copyright 2020 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module Instrumentation
+    module GraphQL
+      # The Instrumentation class contains logic to detect and install the GraphQL
+      # instrumentation
+      class Instrumentation < OpenTelemetry::Instrumentation::Base
+        install do |config|
+          require_dependencies
+          patch_graphql(config)
+        end
+
+        present do
+          defined?(::GraphQL)
+        end
+
+        private
+
+        def require_dependencies
+          require_relative 'patches/opentelemetry_graphql_tracing'
+        end
+
+        def patch_graphql(config = {})
+          config[:schemas].each { |s| patch_schema!(s) }
+        end
```

Comment:
Interestingly, in Shopify's instrumentation, we just:
```ruby
GraphQL::Schema.prepend(Shopify::Tracing::Adapters::GraphQLAdapter::SchemaPatch)
```
where
```ruby
        module SchemaPatch
          def initialize
            super
            @tracers << GraphQLTracer
          end
        end
```
It looks like we're assuming the Schema's `@tracer` is expected to implement `#trace`. I wonder if our implementation predates GraphQL's platform tracing 🤔 .

Reply:
Looks like we're being unnecessarily sleazy here. I believe we can just:
```ruby
GraphQL::Schema.tracer(OpenTelemetry::Instrumentation::GraphQL::Patches::OpenTelemetryGraphQLTracing.new)
```
The [`tracers` method](https://github.com/rmosolgo/graphql-ruby/blob/88a2ebf89e14c7b09b6e9c259b8b4c67ff62d214/lib/graphql/schema.rb#L1578-L1580) collects tracers from up the inheritance hierarchy. I guess the question is: do we want auto-instrumentation to instrument *all* schemas or just specific ones listed by the user?

Reply:
My initial thought here is that we should instrument all the schemas auto-magically, since by and large user's will just be configuring things via the tracer's `c.use_all`. If the goal is "Time to Value",  and simply passing in the otel tracing class to the `.tracer` method permits that, it feels like a big time to value win.  I do have concerns that this approach could be noisy and generate a lot of spans OOTB, which might impact users with storage costs or performance overhead. I'm not sure how much of a real concern that is, but it _seems_ like a thing someone would want to tweak. It might be worth having `schemas` as an optional config setting. If undefined we patch everything, via `GraphQL::Schema.tracer(OpenTelemetry::Instrumentation::GraphQL::Patches::OpenTelemetryGraphQLTracing.new)`. If defined we patch only the schemas specified in the config's `schemas` array.

I'd like to poke around other lang's and vendors here and see how they approach things or if they have surfaced drawbacks to this approach, but generally what do you think about the above?

Reply:
> It might be worth having `schemas` as an optional config setting. If undefined we patch everything, via `GraphQL::Schema.tracer(OpenTelemetry::Instrumentation::GraphQL::Patches::OpenTelemetryGraphQLTracing.new)`. If defined we patch only the schemas specified in the config's `schemas` array.
> 
> I'd like to poke around other lang's and vendors here and see how they approach things or if they have surfaced drawbacks to this approach, but generally what do you think about the above?

Very much 👍